LFS-1115: As an admin, I can look inside installed vocabularies

veronikaslc opened the pull request:
Includes:
[LFS-1118](https://phenotips.atlassian.net/browse/LFS-1118): Ensure all installed vocabularies have an About button whether or not they have a description
[LFS-1116](https://phenotips.atlassian.net/browse/LFS-1116): The vocabulary browser dialog should have the title `<Vocabulary name> (<Vocabulary id>)`
[LFS-1117](https://phenotips.atlassian.net/browse/LFS-1117): Add "Browse" button to the "About" dialog of installed vocabularies

Commit:
LFS-1116: The vocabulary browser dialog should have the title <Vocabulary name> (<Vocabulary id>)

marta- commented on modules/data-entry/src/main/frontend/src/vocabQuery/query.jsx:
```diff
@@ -519,91 +309,190 @@ class VocabularyQuery extends React.Component {
+  if (disabled) {
+    // Alter our text to either the override ("Please select at most X options")
+    // or empty it
+    anchorE.currentl.value = disabled ? props.overrideText : "";
```
Typo

Commit:
LFS-1115: As an admin, I can look inside installed vocabularies
Added roots indicator to the vocabulary term and vocabulary root

Commit:
LFS-1117: Add Browse button to the About dialog of installed vocabularies

sashaandjic commented:
- [ ] LFS-1118: Ensure all installed vocabularies have an About button whether or not they have a description
- [ ] LFS-1116: The vocabulary browser dialog should have the title <Vocabulary name> (<Vocabulary id>)
- [ ] LFS-1117: Add "Browse" button to the "About" dialog of installed vocabularies

**Problem:**
Clicking on BROWSE and LEARN MORE renders just a Vocabulary name and ID.

![image](https://user-images.githubusercontent.com/23243244/116977281-3c305280-ac90-11eb-958c-2f5e4c9bfd5d.png)

![image](https://user-images.githubusercontent.com/23243244/116977261-363a7180-ac90-11eb-9b1c-8a8ca80961b5.png)

Commit:
LFS-1117: Add Browse button to the About dialog of installed vocabularies
Fixed reloading on term change

sashaandjic reviewed (approved):
tested
![image](https://user-images.githubusercontent.com/23243244/117036231-3314a500-acd3-11eb-9260-27c6f640123e.png)

marta- commented:
> **Problem:**
> Clicking on BROWSE and LEARN MORE renders just a Vocabulary name and ID.

I still encounter this issue.

Install HANCESTRO, and as when it's done click on About -> Browse: An empty dialog (just title) pops up.
Reload the page and then click on About -> Browse: A dialog populated with all the roots pops up.

Commit:
LFS-1117: Add Browse button to the About dialog of installed vocabularies
Load vocabulary info after installation when the about info is loaded

veronikaslc commented:
Fixed the bug and rebased on top of LFS-1085 PR.

sashaandjic commented:
I can no longer uninstall vocabulary right after installing it.
![image](https://user-images.githubusercontent.com/23243244/117414211-e2749600-aee4-11eb-8f20-b35ad5c0c437.png)

sashaandjic commented:
Not sure if this is out of scope. of this PR.
When I hit space in Race/Ethnicity field I get this:
![image](https://user-images.githubusercontent.com/23243244/117414788-88c09b80-aee5-11eb-80e9-66e818b07ff3.png)

I would expect to see the same as when pressing on BROWSE button.
![image](https://user-images.githubusercontent.com/23243244/117414909-b279c280-aee5-11eb-8ad1-1cb3afbc8bf8.png)

sashaandjic commented:
Clicking on No sub-categories does not clear tooltips
![image](https://user-images.githubusercontent.com/23243244/117417455-5cf2e500-aee8-11eb-8c4b-b5c1e8c0e888.png)

veronikaslc commented:
I can not uninstull vocabularies for a long time too. This is out of the scope of this PR.
This PR only dealing with vocabulary Admin section and did not alter any vocabulary input code, second issue also out of the scope is it is an issue (the expected scenario is unknown and was never implemented)

sashaandjic commented:
The problem with tooltips not clearing out is only present when clicking on the BROWSE button in the Admin's Vocabulary page.
It cannot be reproduced in the forms as we have vocabulary filters enforced there, so those obsolete categories are not shown.
To reproduce the problem:
- install HANCESTRO
- click on the About button
- click on the BROWSE button
- try to expand any of the obsolete-* categories
 
![image](https://user-images.githubusercontent.com/23243244/117478245-e926fb80-af2c-11eb-81c3-b9d474100605.png)

marta- commented on modules/data-entry/src/main/frontend/src/vocabQuery/browse.jsx:
```diff
@@ -42,14 +42,27 @@ import { REST_URL, MakeRequest } from "./util.jsx";
+    // if we are building for roots
+    if (roots) {
+      let rootBranches = roots.map((row, index) => {
+        return row["identifier"] ? constructBranch(row["identifier"], row["@path"], row["label"], true, false, false, true) : false;
```
Not all roots will have children. This information should be sent by the backend and used here as the last parameter of `constructBranch`.

veronikaslc replied:
Ideally yes, but unfortuantelly we do not index `lfs:hasChildren` property for the terms and it does not get auto-populated to roots when we get vocabulary info.
Also I'd never guess roots can have no children, sounds strange.

marta- replied:
I see. I'll create a separate task for it.

Also we should filter out obsolete terms when querying (see for example `obsolete_Carribean` in the last screenshot).

marta- replied:
Obsoletes are excluded in 84087fb

marta- commented:
> Admin/Vocabularies page's About/Browse - Clicking on No sub-categories does not clear tooltips

Introduced by #629, fixed in 30f7d6a.

sashaandjic reviewed (approved):
tested - all good
![image](https://user-images.githubusercontent.com/23243244/117497462-85a9c780-af46-11eb-9712-e8a20fa6c09c.png)

marta- committed:
LFS-1115: As an admin, I can look inside installed vocabularies
Don't allow focusing on a different term when showing the vocabulary on the Vocabularies admin page

marta- committed:
LFS-1115: As an admin, I can look inside installed vocabularies
Fixed regression: clicking on (i) in the vocabulary browser on the Vocabularies admin page results in white screen and console error.
Introduced in the last rebase which changed the `InfoBox` props.

marta- reviewed (dismissed):
Bugs:
* vocab admin page: focusing on a term does not work
* vocab admin page:clicking on (i) results in white screen and console error
* patient info form > race/ethnicity: search for a term, click on (i) to open the info box, click on "Learn more" to open the browser => info box stays above vocabulary browser

marta- commented:
> Bugs:
> 
> * vocab admin page: focusing on a term does not work

Addressed 9482f86 and c8019e6. Instead of showing a loading sign with no effect, don't allow focusing on a different term while on the admin page, and expand/collapse on click instead.

> * vocab admin page:clicking on (i) results in white screen and console error

Fixed in 663679f.

> * patient info form > race/ethnicity: search for a term, click on (i) to open the info box, click on "Learn more" to open the browser => info box stays above vocabulary browser

@veronikaslc to do

marta- commented:
![image](https://user-images.githubusercontent.com/651980/117510439-150ca600-af5a-11eb-8159-f1b6cf2fac74.png)

`CC BY 4.0` should not be here, that's the license. @veronikaslc please investigate if it's an issue with indexing (in which case it's out off scope) or with how we obtain roots (in which case it's within scope of this PR).

marta- committed:
LFS-1115: As an admin, I can look inside installed vocabularies
Exclude obsolete terms when returning the roots of a vocabulary

veronikaslc commented:
>patient info form > race/ethnicity: search for a term, click on (i) to open the info box, click on "Learn more" to open the browser => info box stays above vocabulary browser
Out of the scope, was introduced [here](https://github.com/ccmbioinfo/lfs/pull/627/commits/86e39703aee15544861a987de697278ab87954e8) the change was requested in one of comments to the [another PR](https://github.com/ccmbioinfo/lfs/pull/627)

marta- committed:
LFS-1115: As an admin, I can look inside installed vocabularies
Exclude obsolete terms when returning the roots of a vocabulary -- refactor to fix checkstyle errors

marta- commented:
> Last commit breaks the build on some checkstyle errors.

Codestyle fixed in e5ba10f .

marta- commented:
> > > patient info form > race/ethnicity: search for a term, click on (i) to open the info box, click on "Learn more" to open the browser => info box stays above vocabulary browser
> > 
> > 
> > Out of the scope, was introduced [here](https://github.com/ccmbioinfo/lfs/pull/627/commits/86e39703aee15544861a987de697278ab87954e8) the change was requested in one of comments to the [another PR](https://github.com/ccmbioinfo/lfs/pull/627)
> 
> I was checking both 1115 and 1085 in parallel and could not reproduce this issue in 1085.

@veronikaslc My mistake, I confirm it's present in 1085.

Commit:
LFS-1115: As an admin, I can look inside installed vocabularies
Rebased on the lastest refactoring of LFS-1085

marta- commented on modules/vocabularies/src/main/frontend/src/vocabularyDetails.jsx:
```diff
@@ -60,16 +63,32 @@ const useStyles = makeStyles(theme => ({
+  browseAction: {
+    margin: theme.spacing(1),
+    textTransform: "none",
+    marginRight: theme.spacing(37)
```
I don't think we need a magic number, let's see if this works:
```suggestion
    marginRight: "auto"
```

marta- replied:
Done in 33f5307 .

marta- committed:
LFS-1115: As an admin, I can look inside installed vocabularies
Updated imported component name after rebase

marta- committed:
LFS-1115/LFS-1117: Add "Browse" button to the "About" dialog of installed vocabularies
Simplify styling

marta- committed:
LFS-1115: As an admin, I can look inside installed vocabularies
Bugfix: the browser is empty right after installing the vocabulary

marta- committed:
LFS-1115: As an admin, I can look inside installed vocabularies
Improvement: if there's only one root, display it expanded

marta- reviewed (commented):
Issues:
* Build fails because the renaming of `InfoBrowser` to `VocabularyBrowser` was not accounted for after the last rebase
* Once the right component is imported and used, the browser is empty if accessed immediately after installing the vocabulary (but works after refreshing the page)

Improvement:
* If there's only one root, it should be displayed expanded

marta- commented:
Rebased on the latest 1085.

> * Build fails because the renaming of `InfoBrowser` to `VocabularyBrowser` was not accounted for after the last rebase

Fixed in 5e91098 .

> * Once the right component is imported and used, the browser id empty if accessed immediately after installing the vocabulary (works after refreshing the page)

Fixed in 59cfbb7 .

> Improvement:
> 
> * If there's only one root, it should be displayed expanded

Done in 2ef955c .

marta- commented:
Usability issue:
* On the admin/Vocabularies page, install HANCESTRO (or any other vocabulary with more than one root) and open the browser for it
* Expand a root, then click on one of the descendants to "focus" the browser on it

Now there's no way to go back to seeing all the roots unless you close the browser and reopen it.

I see two options:
1. Don't focus on click. This was implemented in one of the earlier versions of this branch, when clicking would just expand the term instead of focusing on it. Pro: the whole vocabulary is always on the screen, no confusion about where you are or how to get back. Con: might become extremely sluggish for large vocabularies.
2. Focus on click, but always add a `...` button allowing to focus on the parent, even if the parent is null. For a null parent, `...` would have no label and clicking on it would list all the roots.

Thoughts? @sashaandjic @veronikaslc

sashaandjic commented:
> I see two options:
> 
> 1. Don't focus on click. This was implemented in one of the earlier versions of this branch, when clicking would just expand the term instead of focusing on it. Pro: the whole vocabulary is always on the screen, no confusion about where you are or how to get back. Con: might become extremely sluggish for large vocabularies.
> 2. Focus on click, but always add a `...` button allowing to focus on the parent, even if the parent is null. For a null parent, `...` would have no label and clicking on it would list all the roots.

If it is not too much of a work, I'd like to see how bad option 1 is.

marta- commented:
> If it is not too much of a work, I'd like to see how bad option 1 is with a large vocabulary.

@sashaandjic Focus disabled in the latest commit (ca404e8), please give it a try.

marta- committed:
LFS-1115: As an admin, I can look inside installed vocabularies
Disabled term focus when browsing vocabulary from the root(s), to avoid usability issue that prevents user from going back to all the roots after focusing on another term

Commit:
LFS-1115: As an admin, I can look inside installed vocabularies
Fixed info box click away functionality to close the box on browser interaction

sashaandjic commented:
Should BROWSER button be available for installed vocabularies in the Bioportal's list at the bottom?

![image](https://user-images.githubusercontent.com/23243244/118726414-48bfb980-b7ff-11eb-8231-7adf5cd9db2f.png)

![image](https://user-images.githubusercontent.com/23243244/118726359-30e83580-b7ff-11eb-9ba4-7e5f50f082c8.png)

marta- commented:
> Should BROWSER button be available for installed vocabularies in the Bioportal's list at the bottom?

That would be nice.

sashaandjic commented:
Sometimes I experience this jump as shown in the video. I clicked on fermion to collapse it, and it went all the way down in the list:

https://user-images.githubusercontent.com/23243244/118727029-24181180-b800-11eb-88b9-193cba541b8d.mp4

sashaandjic commented:
Having ability to open/close info box by typing on "i" button would be beneficial when browsing through a large list of terms. I would keep existing functionality too (closing info box by pressing either on "x" or by clicking outside of the box)

marta- commented:
> Sometimes I experience this jump as shown in the video. I clicked on `fermion `to collapse it, and it went all the way down in the list:

That's because of the scroll position. I don't think there's much we can do about it. We may want to consider a brief highlight (couple of seconds) of the last term clicked on.

sashaandjic commented:
All in all this PR looks good. The added benefit would be to have:
- BROWSE bttn in BioPortal list
- toggle on "i" 

However, it can be merged as is.

marta- commented:
> All in all this PR looks good. The added benefit would be to have:
> 
> * BROWSE bttn in BioPortal list

This is probably easy to do and we should have it.

> * toggle on "i"

This is for @veronikaslc to decide if it's worth the struggle. I'm fine either way.

Commit:
LFS-1115: As an admin, I can look inside installed vocabularies
Fixed icon button functionality

marta- commented on modules/data-entry/src/main/frontend/src/vocabQuery/VocabularyQuery.jsx:
```diff
@@ -244,7 +244,7 @@ function VocabularyQuery(props) {
     }
 
-    setInputValue("");
+    !anchorEl.current.contains(event.target) && setInputValue("");
```
`!anchorEl?.current?.contains`

sashaandjic commented:
I think the latest change broke the closing of an info box if clicking on a different term. I did not notice this behaviour previously.

sashaandjic commented:
@veronikaslc We should add BROWSE bttn to the installed vocabularies in the BioPortal grid

Commit:
LFS-1115/LFS-1117: Add Browse button to the About dialog of installed vocabularies
Show Browse button for both tables for installed vocabs

veronikaslc commented:
@sashaandjic Just added

Commit:
LFS-1115: As an admin, I can look inside installed vocabularies
Close info box on the term click

veronikaslc commented:
@sashaandjic Fixed the the closing of an info box if clicking on a term.

marta- commented on modules/data-entry/src/main/frontend/src/vocabQuery/VocabularyBranch.jsx:
```diff
@@ -67,6 +68,7 @@ function VocabularyBranch(props) {
       onTermClick(path);
     } else {
       toggleShowChildren();
+      onCloseInfoBox();
```
Either mark as required in PropTypes or check here if it's defined

sashaandjic reviewed (approved):
tested

veronikaslc committed:
LFS-1115: As an admin, I can look inside installed 

Addressed codereview comments

sashaandjic reviewed (approved):
re-tested after the latest change

marta- commented on modules/vocabularies/src/main/frontend/src/vocabularyDetails.jsx:
```diff
@@ -109,15 +110,14 @@ export default function VocabularyDetails(props) {
 
         <DialogActions>
-          { type == "local" &&
+          {(phase == Phase["Latest"] || phase == Phase["Update Available"]) && 
```
These changes appear to generate attempts to load local vocabulary data for all vocabularies as the Bioportal table, resulting in a lot of unnecessary server calls, all 404. Can we check client side if the vocabulary is in the `local` list instead of making a server request?
![image](https://user-images.githubusercontent.com/651980/118996292-c7c70600-b955-11eb-951f-0b33f00d03b6.png)

marta- commented on modules/data-entry/src/main/frontend/src/vocabQuery/VocabularyBrowser.jsx:
```diff
@@ -167,20 +179,24 @@ function VocabularyBrowser(props) {
           registerInfo={registerInfoButton}
           getInfo={getInfo}
+          browseRoots={browseRoots}
+          roots={vocab.roots}
```
Why is this param necessary? Can't we just read it from `vocab`?

Commit:
LFS-1115: As an admin, I can look inside installed vocabularies
Removed unused param; Prevent browser from default rendering

sashaandjic reviewed (approved):
tested

marta- commented on modules/data-entry/src/main/frontend/src/vocabQuery/VocabularyTree.jsx:
```diff
@@ -124,12 +156,12 @@ function VocabularyTree(props) {
       </DialogTitle>
+      <IconButton aria-label="close" className={classes.closeButton} onClick={onClose}>
+        <CloseIcon />
+      </IconButton>
```
After rebasing on the latest dev, the `<Dialog>` can be replaced with the new `ResponsiveDialog` from `commons/components/ResponsiveDialog.jsx`:

```
    <ResponsiveDialog
      title={`${vocabulary.name} (${vocabulary.acronym})` || "Related terms"}
      withCloseButton
      open={open}
      ref={browserRef}
      onClose={onClose}
      classes={{
        paper: classes.dialogPaper,
        root: classes.infoDialog
      }}
      {...rest}
    >
      <DialogContent className={classes.treeContainer} dividers>
        ...
      </DialogContent>
    </ResponsiveDialog>
```
and the `<DialogTitle>` and `<IconButton>` can be deleted along with their styling.

veronikaslc replied:
Function components cannot be given refs.
If we want to use it here like that we have to redefine `ResponsiveDialog` component with `React.forwardRef` call. (done in 78a2117)

marta- commented on modules/data-entry/src/main/frontend/src/vocabQuery/browseStyle.jsx:
```diff
@@ -31,9 +31,12 @@ const browseStyle = theme => ({
     },
     closeButton: {
       position: 'absolute',
-      right: theme.spacing(1),
+      right: '0',
       top: theme.spacing(1),
     },
+    browserTitle: {
+      paddingRight: theme.spacing(6)
+    },
```
Are these two classes still used anywhere?

marta- reviewed (changes requested):
Commit e47bbf8 appears to have caused a regression: when clicking on a term to focus it (in the form, not on the admin page), the term id displayed collapsed. It should be expanded and all its children should be displayed.

veronikaslc commented:
> Commit [e47bbf8](https://github.com/ccmbioinfo/lfs/commit/e47bbf8e234f36acd67851ae238222f719619694) appears to have caused a regression: when clicking on a term to focus it (in the form, not on the admin page), the term id displayed collapsed. It should be expanded and all its children should be displayed.

Fixed in 204013a.

sashaandjic commented:
Browsing isn't working on the Admin's Vocabulary page for locally installed vocabularies. 
![image](https://user-images.githubusercontent.com/23243244/119692844-c1370380-be19-11eb-862a-1f70e8df59e1.png)

No problems were observed in the forms.
![image](https://user-images.githubusercontent.com/23243244/119692922-d318a680-be19-11eb-8e53-7ffbf486256e.png)

To reproduce, install HANCESTRO from : https://raw.githubusercontent.com/EBISPOT/ancestro/master/hancestro_inferred.owl

marta- commented:
> Browsing isn't working on the Admin's Vocabulary page for locally installed vocabularies.

Confirmed. [FileUploadIndexer.java](https://github.com/ccmbioinfo/lfs/blob/2311f2721b7bd6f9cd39233284df2a2bdbff035b/modules/vocabularies/src/main/java/ca/sickkids/ccm/lfs/vocabularies/internal/FileUploadIndexer.java) should index the root nodes in the same way [BioOntologyIndexer.java](https://github.com/ccmbioinfo/lfs/pull/626/files#diff-0e11cd485f102c8cf7bffb25fdf8a8f268740f564117d823dc3e048ce5661497R134) does.

marta- commented:
> > Browsing isn't working on the Admin's Vocabulary page for locally installed vocabularies.
> 
> Confirmed. [FileUploadIndexer.java](https://github.com/ccmbioinfo/lfs/blob/2311f2721b7bd6f9cd39233284df2a2bdbff035b/modules/vocabularies/src/main/java/ca/sickkids/ccm/lfs/vocabularies/internal/FileUploadIndexer.java) should index the root nodes in the same way [BioOntologyIndexer.java](https://github.com/ccmbioinfo/lfs/pull/626/files#diff-0e11cd485f102c8cf7bffb25fdf8a8f268740f564117d823dc3e048ce5661497R134) does.

Fixed in 17e2cee

marta- committed:
LFS-1115: As an admin, I can look inside installed vocabularies
Bugfix: clicking Browse for a vocabulary installed from a local file always displays an empty browser

sashaandjic commented:
Clicking outside of the browser window does not close it. The only way to clos it is to click on "x" button
![image](https://user-images.githubusercontent.com/23243244/119716615-4844a580-be33-11eb-8306-4f219157bf20.png)

sashaandjic commented:
Unable to uninstall vocabulary even if it was not used at all in the form.
![image](https://user-images.githubusercontent.com/23243244/119716720-69a59180-be33-11eb-8ce6-e19bcc913b44.png)

sashaandjic commented:
Probably out of scope: There is no distinction between locally installed vocabulary and one from the Bioportal.In the screenshot below, I installed HANCESTR form the local file. HANCESTRO from Bioportal seems to be show as installed too, even though it was not.
![image](https://user-images.githubusercontent.com/23243244/119716960-b7ba9500-be33-11eb-8ead-3f4fb502bd14.png)

marta- commented:
> Clicking outside of the browser window does not close it. The only way to clos it is to click on "x" button

That's intentional, IMO an accidental click outside should not make the results of a lot of digging through the ontology disappear. Is  this considered a regression, should we change it back?

> Unable to uninstall vocabulary even if it was not used at all in the form.

Out of scope, has been reported as https://phenotips.atlassian.net/browse/LFS-1150 .

> Probably out of scope: There is no distinction between locally installed vocabulary and one from the Bioportal.In the screenshot below, I installed HANCESTR form the local file. HANCESTRO from Bioportal seems to be show as installed too, even though it was not.

We need to discuss whether (or how) we want to present such a distinction. For now we identify a vocabulary by its id, regardless of where its source came from. Please create a task to open the discussion and come up with a plan.

sashaandjic commented:
> That's intentional, IMO an accidental click outside should not make the results of a lot of digging through the ontology disappear. Is this considered a regression, should we change it back?

Let's keep it this way.

sashaandjic reviewed (approved):
tested

Commit:
LFS-1115: As an admin, I can look inside installed vocabularies
Separated callbacks for info box and tree close events

veronikaslc commented:
> Clicking outside of the browser window does not close it. The only way to clos it is to click on "x" button

Change was requested by @sdumitriu not ot loose the browser navigation experience.

> Unable to uninstall vocabulary even if it was not used at all in the form.

Was already found and reported in this PR 20 days ago. It seems we started going circles here since both 1085 and 1115 are big. May be keeping a summary of all that was found/resolved could help?

> Probably out of scope: There is no distinction between locally installed vocabulary and one from the Bioportal.In the screenshot below, I installed HANCESTR form the local file. HANCESTRO from Bioportal seems to be show as installed too, even though it was not.

Now we do not index the source nature of the vocabulary. Might be worth to do so.

sashaandjic reviewed (approved):
tested

sashaandjic commented:
Can we move down vocabulary browser so it covers the About in Admin's Vocabularies page?
![image](https://user-images.githubusercontent.com/23243244/119953932-7da0de80-bf6c-11eb-982e-2f87b67eb985.png)
![image](https://user-images.githubusercontent.com/23243244/119954002-8f828180-bf6c-11eb-9fe0-b35c7de57ea6.png)

marta- commented:
> **QUESTION**: Can we move down vocabulary browser so it covers the About in Admin's Vocabularies page?

We can, but when I tried it it looked really bad on expanding and collapsing terms, because the height will vary and the top of the dialog will keep jumping up and down as you navigate the vocabulary.

sashaandjic reviewed (approved):
tested

sdumitriu commented on modules/vocabularies/src/main/java/ca/sickkids/ccm/lfs/vocabularies/internal/OntologyIndexerUtils.java:
```diff
@@ -46,6 +48,9 @@
     /** The list which holds all JCR vocabulary nodes associated with a vocabulary to be checked-in. */
     private static final ThreadLocal<List<Node>> NODES_TO_CHECK_IN = ThreadLocal.withInitial(ArrayList::new);
 
+    /** The list which holds all root vocabulary nodes to add to vocabulary node. */
```
This sounds a bit awkward.

```suggestion
    /** The list which holds all root terms. */
```

sdumitriu commented on modules/vocabularies/src/main/java/ca/sickkids/ccm/lfs/vocabularies/internal/FileUploadIndexer.java:
```diff
@@ -147,6 +147,8 @@ public void index(final String source, final SlingHttpServletRequest request,
             // Parse the source file and create VocabularyTerm node children
             parser.parse(temporaryFile, description, this::createVocabularyTermNode);
 
+            this.vocabularyNode.get().setProperty("roots", OntologyIndexerUtils.getRootNodes());
```
This should be done in `OntologyIndexerUtils.finalizeInstall`, there's no need to expose `getRootNodes`, or duplicate the code.

sashaandjic commented:
Bug found when maxAnswers is not 0.
- in Patient Information form's `Race/Ethnicity` field set `maxAnswers `to any number other than 0. (e.g. `maxAnswers = 2`)
- create Patient Information form and select max number of options
- save and close
- try to open the form in edit mode --> white screen of death appears

![image](https://user-images.githubusercontent.com/23243244/120562805-24560680-c3d5-11eb-910f-69132e070106.png)

veronikaslc commented:
@sashaandjic Out of the scope of this PR

veronikaslc committed:
LFS-1115: As an admin, I can look inside installed vocabularies

Improved function description

Co-authored-by: Sergiu Dumitriu <sdumitriu@users.noreply.github.com>

Commit:
LFS-1115: As an admin, I can look inside installed
Optimised root recording

sashaandjic commented:
@Veronika I can always  reproduce the above problem in this branch, not in dev.

Step to reproduce:
- set maxAnswers=2 for Patien Information form
- create a form and select 2 nationalities
- save and close
- open in edit mode

https://user-images.githubusercontent.com/23243244/120727123-bcbcbb80-c4a7-11eb-90d6-b0e9439200f6.mp4

veronikaslc commented:
@sashaandjic I see. I had tested any number other than 0 - tried 1 - and it worked so I could not reproduce it. Appeared its any number above 1.

Commit:
LFS-1085: Refactor query.jsx as a functional component; LFS-1151: Show no results when typing space in the vocabulary field
Fixed absent anchor error

sashaandjic commented:
The old bug is back - unable to uninstall vocabularies including ones that are not in use at all like OGROUP
![image](https://user-images.githubusercontent.com/23243244/120752548-e4c41300-c4d7-11eb-8d41-1087b848391a.png)
![image](https://user-images.githubusercontent.com/23243244/120752564-ebeb2100-c4d7-11eb-9063-3f8930bbbb47.png)

sashaandjic commented:
> @sashaandjic I see. I had tested any number other than 0 - tried 1 - and it worked so I could not reproduce it. Appeared its any number above 1.

@veronika- ah I did not realize it was affecting `maxAnswers>1` not `maxAnswers>0` and I just realized that this bug was also present in dev - I tested dev by using maxAnswer=1 and wrongly concluded that bug was not present in the dev branch. SORRY! 

The new [task ](https://phenotips.atlassian.net/browse/LFS-1185) is created - please add it to this PR - I tested the fix and it works

veronikaslc commented:
>The old bug is back 
@sashaandjic We have a separate task for it - LFS-1150. That is out of the scope of this PR. This PR **only addresses LFS-1115**, LFS-1150 was never a part of this PR and was never fixed here. LFS-1150 is fixed and merged to dev separately, before LFS-1115. Also LFS-1115 was never based on LFS-1150 and never included that fix.

veronikaslc commented:
I will rebase the PR on the latest dev to include all the latest fixed to avoid the confusion.

Note: This PR does not fix reported issues for Filters or vocabulary answers in Forms. It practically bypasses all of the code for Filters and vocabulary inputs. Also it does not fix any bugs or alter installation/uninstallation process.

Commit:
LFS-1150: LFS-1150: Unable to uninstall vocabularies; LFS-1180: Add vocabulary name to the uninstall warning dialog
Fixed term path comparison

sashaandjic reviewed (approved):
Tested

- [x] LFS-1118: Ensure all installed vocabularies have an About button whether or not they have a description
- [x] LFS-1116: The vocabulary browser dialog should have the title <Vocabulary name> (<Vocabulary id>)
- [x] LFS-1117: Add "Browse" button to the "About" dialog of installed vocabularies
- [x] LFS-1185: Unable to open form in edit mode once max answers are reached and `maxAnswer>1`

sdumitriu committed:
LFS-1150: Unable to uninstall vocabularies
Better detection of self-contained references
